[SPARK-39203][SQL] Rewrite table location to absolute URI based on database URI

wangyum opened the pull request:
### What changes were proposed in this pull request?

Rewrite table location to absolute location based on database location. For example:
```
Table location: /user/hive/warehouse/db_spark_39203.db/t1
Database location: viewfs://clusterA/user/hive/warehouse/db_spark_39203.db/
```
The current table location is: `viewfs://clusterA/user/hive/warehouse/db_spark_39203.db/t1`.

### Why are the changes needed?

The old Spark version(before SPARK-19257) will not use absolute path when using the following SQL create table:
```sql
CREATE TABLE DB_SPARK_39203.t6(id int)
USING parquet
OPTIONS (
  path '/user/hive/warehouse/db_spark_39203.db/t1'
);
```
<img width="551" alt="image" src="https://user-images.githubusercontent.com/5399861/170249177-dded6f23-2a59-4d08-a410-df4c975693b7.png">


This issue makes Spark can't read tables across cluster. For example:
Hive, Spark 2.1 and HDFS 1 on cluster A.
Spark 3.0 and HDFS 2 on cluster B.

The other create table command and latest spark do not have this issue:
```sql
CREATE TABLE DB_SPARK_39203.t7(id int) using parquet;

CREATE TABLE DB_SPARK_39203.t8(id int)
stored as parquet
LOCATION '/user/hive/warehouse/db_spark_39203.db/t1';

CREATE EXTERNAL TABLE DB_SPARK_39203.t9(id int)
stored as parquet
LOCATION '/user/hive/warehouse/db_spark_39203.db/t1';
```
<img width="569" alt="image" src="https://user-images.githubusercontent.com/5399861/170249397-8d1677bc-b506-4d66-b703-928367c9d780.png">



### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test and production test: we have been using this patch for over two years.

wangyum committed:
fix

wangyum committed:
SPARK-39203: Rewrite table location to absolute location based on database location

wangyum committed:
Simplify

wangyum commented:
@cloud-fan

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala:
```diff
@@ -842,10 +843,12 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
-      val tableLocation = getLocationFromStorageProps(table)
+      val tableLocation = getLocationFromStorageProps(table).map { loc =>
+        val uri = CatalogUtils.stringToURI(loc)
+        table.storage.locationUri.map(HiveExternalCatalog.toAbsoluteURI(uri, _)).getOrElse(uri)
```
`table.storage.locationUri` is the table location, not database location, isn't it?

wangyum replied:
The table location uri has been rewritten by HiveClientImpl.convertHiveTableToCatalogTable.
For example:
```
table.storage.locationUri: Some(viewfs://clusterA/user/hive/warehouse/db_spark_39203.db/t1)
loc: /user/hive/warehouse/db_spark_39203.db/t1
```

cloud-fan replied:
Sorry I don't understand. The previous code completely ignores `table.storage.locationUri`. Do you mean we still don't care about `table.storage.locationUri`, but only want to get its qualifier (scheme, host, port, etc.)?

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:
```diff
@@ -518,7 +518,15 @@ private[hive] class HiveClientImpl(
+        locationUri = shim.getDataLocation(h).map { location =>
+          val tableUri = CatalogUtils.stringToURI(location)
+          if (!tableUri.isAbsolute) {
+            val dbUri = CatalogUtils.stringToURI(client.getDatabase(h.getDbName).getLocationUri)
```
It seems to be that it's clearer to get database location in `HiveExternalCatalog.restoreDataSourceTable` and do not change `HiveClientImpl` at all.

wangyum replied:
It can't handle all cases. For example `rawTable`'s location is incorrect:
https://github.com/apache/spark/blob/4df8512b11dc9cc3a179fd5ccedf91af1f3fc6ee/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala#L1283-L1289

https://github.com/apache/spark/blob/4df8512b11dc9cc3a179fd5ccedf91af1f3fc6ee/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala#L761-L769

This makes `fromHivePartition` can' correct the partition location:
https://github.com/apache/spark/blob/4df8512b11dc9cc3a179fd5ccedf91af1f3fc6ee/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala#L1157

wangyum committed:
Fix

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:
```diff
@@ -518,7 +519,15 @@ private[hive] class HiveClientImpl(
-        locationUri = shim.getDataLocation(h).map(CatalogUtils.stringToURI),
+        locationUri = shim.getDataLocation(h).map { loc =>
+          val tableUri = CatalogUtils.stringToURI(loc)
+          val parentUri = if (tableUri.isAbsolute) {
```
for tables created by recent versions of Spark, this condition is always true?

wangyum replied:
Yes.

wangyum replied:
<img width="582" alt="image" src="https://user-images.githubusercontent.com/5399861/170652237-ab9d2fc9-e9ce-4b34-be76-8e0ca8814722.png">
and the other command:

```
spark-sql>
         > CREATE TABLE DB_SPARK_39203.t7(id int) using parquet;
Time taken: 0.43 seconds
spark-sql>
         > CREATE TABLE DB_SPARK_39203.t8(id int)
         > stored as parquet
         > LOCATION '/user/hive/warehouse/db_spark_39203.db/t1';
Time taken: 0.265 seconds
spark-sql>
         > CREATE EXTERNAL TABLE DB_SPARK_39203.t9(id int)
         > stored as parquet
         > LOCATION '/user/hive/warehouse/db_spark_39203.db/t1';
Time taken: 0.278 seconds
spark-sql> desc formatted DB_SPARK_39203.t7;
id	int	NULL

# Detailed Table Information
Database	db_spark_39203
Table	t7
Owner	####
Created Time	Fri May 27 00:31:09 GMT-07:00 2022
Last Access	UNKNOWN
Created By	Spark 3.2.0-SNAPSHOT
Type	MANAGED
Provider	parquet
Location	viewfs://####/user/hive/warehouse/db_spark_39203.db/t7
Serde Library	org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe
InputFormat	org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat
OutputFormat	org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat
Time taken: 0.152 seconds, Fetched 15 row(s)
spark-sql> desc formatted DB_SPARK_39203.t8;
id	int	NULL

# Detailed Table Information
Database	db_spark_39203
Table	t8
Owner	####
Created Time	Fri May 27 00:31:09 GMT-07:00 2022
Last Access	UNKNOWN
Created By	Spark 3.2.0-SNAPSHOT
Type	EXTERNAL
Provider	hive
Table Properties	[transient_lastDdlTime=1653636669]
Location	viewfs://####/user/hive/warehouse/db_spark_39203.db/t1
Serde Library	org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe
InputFormat	org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat
OutputFormat	org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat
Storage Properties	[serialization.format=1]
Partition Provider	Catalog
Time taken: 0.224 seconds, Fetched 18 row(s)
spark-sql> desc formatted DB_SPARK_39203.t9;
id	int	NULL

# Detailed Table Information
Database	db_spark_39203
Table	t9
Owner	####
Created Time	Fri May 27 00:31:10 GMT-07:00 2022
Last Access	UNKNOWN
Created By	Spark 3.2.0-SNAPSHOT
Type	EXTERNAL
Provider	hive
Table Properties	[transient_lastDdlTime=1653636670]
Location	viewfs://####/user/hive/warehouse/db_spark_39203.db/t1
Serde Library	org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe
InputFormat	org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat
OutputFormat	org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat
Storage Properties	[serialization.format=1]
Partition Provider	Catalog
Time taken: 0.142 seconds, Fetched 18 row(s)
```

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:
```diff
@@ -753,7 +762,13 @@ private[hive] class HiveClientImpl(
+    val parentUri = if (hiveTable.getDataLocation.toUri.isAbsolute) {
+      None
+    } else {
+      convertHiveTableToCatalogTable(hiveTable).storage.locationUri
```
shoudn't this be `CatalogUtils.stringToURI(client.getDatabase(h.getDbName).getLocationUri)`?

wangyum replied:
+1

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala:
```diff
@@ -1432,4 +1434,24 @@ object HiveExternalCatalog {
+   *   The result is: viewfs://clusterA/user/hive/warehouse/test_table
+   */
+  private[spark] def toAbsoluteURI(uri: URI, parentUri: Option[URI]): URI = {
+    if (uri.isAbsolute) {
```
nit:
```
if (!uri.isAbsolute && parentUri.isDefined) {
  new URI...
} else {
  uri
}
```

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala:
```diff
@@ -1432,4 +1434,24 @@ object HiveExternalCatalog {
+   *   parentUri: viewfs://clusterA/user/hive/warehouse/
+   *   The result is: viewfs://clusterA/user/hive/warehouse/test_table
+   */
+  private[spark] def toAbsoluteURI(uri: URI, parentUri: Option[URI]): URI = {
```
I think the name `parentUri` is a bit misleading. It's more like an `absoluteUri` which we need to inherit its schema, host, port, etc.

wangyum replied:
Renamed it to absoluteUri.

wangyum committed:
Address comments

wangyum committed:
fix:
[info] - 0.12: getPartitionNames(catalogTable) (74 milliseconds)
[info] org.apache.spark.sql.hive.client.HiveClientSuites *** ABORTED *** (13 seconds, 434 milliseconds)
[info]   java.lang.NoSuchMethodError: org.apache.hadoop.hive.ql.metadata.Table.getDataLocation()Lorg/apache/hadoop/fs/Path;
[info]   at org.apache.spark.sql.hive.client.HiveClientImpl.$anonfun$getPartitions$3(HiveClientImpl.scala:765)

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:
```diff
@@ -518,7 +520,12 @@ private[hive] class HiveClientImpl(
       storage = CatalogStorageFormat(
-        locationUri = shim.getDataLocation(h).map(CatalogUtils.stringToURI),
+        locationUri = shim.getDataLocation(h).map { loc =>
+          val tableUri = stringToURI(loc)
```
can we add some code comments to explain this backward compatibility story?

wangyum replied:
OK

wangyum committed:
fix

wangyum commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:
```diff
@@ -753,7 +760,10 @@ private[hive] class HiveClientImpl(
+    val absoluteUri = shim.getDataLocation(hiveTable).map(stringToURI).filter(!_.isAbsolute)
+      .map(_ => stringToURI(client.getDatabase(hiveTable.getDbName).getLocationUri))
+    val parts = shim.getPartitions(client, hiveTable, partSpec.asJava)
+      .map(fromHivePartition(_, absoluteUri))
```
No. the `absoluteUri ` is None if table location is absolute uri.

wangyum commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala:
```diff
@@ -842,10 +843,11 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
     val storageWithLocation = {
-      val tableLocation = getLocationFromStorageProps(table)
+      val tableLocation = getLocationFromStorageProps(table).map { path =>
+        toAbsoluteURI(CatalogUtils.stringToURI(path), table.storage.locationUri)
```
OK
